pkg/operator/ceph/cluster/osd/osd.go

```diff
@@ -785,13 +508,59 @@ func (c *Cluster) getOSDInfo(d *apps.Deployment) ([]OSDInfo, error) {
 	}
 
 	if osd.UUID == "" || osd.BlockPath == "" {
-		return []OSDInfo{}, errors.Errorf("failed to get required osdInfo. %+v", osd)
+		return OSDInfo{}, errors.Errorf("failed to get required osdInfo. %+v", osd)
 	}
 
-	return []OSDInfo{osd}, nil
+	return osd, nil
+}
+
+func osdIsOnPVC(d *appsv1.Deployment) bool {
+	if _, ok := d.Labels[OSDOverPVCLabelKey]; ok {
+		return true
+	}
+	return false
 }
 
-func getLocationFromPod(clientset kubernetes.Interface, d *apps.Deployment, crushRoot string) (string, string, error) {
+func getNodeOrPVCName(d *appsv1.Deployment) (string, error) {
+	if v, ok := d.Labels[OSDOverPVCLabelKey]; ok {
+		return v, nil // OSD is on PVC
+	}
+	for k, v := range d.Spec.Template.Spec.NodeSelector {
+		if k == corev1.LabelHostname {
+			return v, nil
+		}
+	}
+	return "", fmt.Errorf("failed to find node/PVC name for OSD deployment %q: %+v", d.Name, d)
+}
+
+// Needed for upgrades from v1.5 to v1.6
+func getBlockPathFromActivateInitContainer(d *appsv1.Deployment) (string, error) {
+	initContainers := d.Spec.Template.Spec.InitContainers
+	for _, c := range initContainers {
+		if c.Name != activatePVCOSDInitContainer {
+			continue
+		}
+		if len(c.Command) != 3 {
+			return "", fmt.Errorf("activate init container has fewer command arguments (%d) than expected (3)", len(c.Command))
+		}
+		script := c.Command[2]
+		varAssignment := "DEVICE=" // this variable assignment is followed by the block path
+		scanner := bufio.NewScanner(strings.NewReader(script))
+		for scanner.Scan() {
```

Comment:
hmm, we really have to scan the init container script to extract this?

Reply:
Yes. There is no other place where it is stored on legacy deployments.